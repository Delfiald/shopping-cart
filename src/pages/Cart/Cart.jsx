import { useOutletContext } from "react-router-dom";
import CartMain from "../../components/Main/CartMain";

function Cart() {
 const { cartItem, setCartItem, wishlistItem, setWishlistItem } =
  useOutletContext();
 return (
  <>
   <CartMain
    cartItem={cartItem}
    setCartItem={setCartItem}
    wishlistItem={wishlistItem}
    setWishlistItem={setWishlistItem}
   />
  </>
 );
}

export default Cart;
