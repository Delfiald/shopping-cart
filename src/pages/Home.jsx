function Home() {
 return (
  <>
   <header>Shoppers</header>
   <main>Main</main>
   <footer>Footer</footer>
  </>
 );
}

export default Home;
