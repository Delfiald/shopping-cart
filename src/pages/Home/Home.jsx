import Main from "../../components/Main/HomeMain";
import { useOutletContext } from "react-router-dom";

function Home() {
 const { categories, products, isExiting, setIsExiting } = useOutletContext();

 return (
  <>
   <Main
    categories={categories}
    products={products}
    isExiting={isExiting}
    setIsExiting={setIsExiting}
   />
  </>
 );
}

export default Home;
