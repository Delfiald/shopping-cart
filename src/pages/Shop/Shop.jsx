import ShopMain from "../../components/Main/ShopMain";
import Aside from "../../components/Aside/Aside";
import { useOutletContext, useSearchParams } from "react-router-dom";
import { useEffect } from "react";

function Shop() {
 const [searchParams, setSearchParams] = useSearchParams();
 const category = searchParams.get("category");
 const { categories, products, hoverButton, setHoverButton } =
  useOutletContext();

 const page = searchParams.get("page");
 const itemPerPage = searchParams.get("itemsPerPage");
 const sort = searchParams.get("sort");
 const search = searchParams.get("search") || "";

 const displayedProducts = products
  .filter((product) => {
   if (category && category !== "all") {
    return product.category === category;
   }
   return true;
  })
  .filter((product) => {
   if (search) {
    return product.title
     .toLowerCase()
     .replace(/\s+/g, "")
     .includes(search.toLowerCase());
   }
   return true;
  });

 const handleFilterChange = (newCategory) => {
  setSearchParams((prevParams) => {
   const params = new URLSearchParams(prevParams);
   params.set("category", newCategory);
   params.set("page", 1);
   return params;
  });
 };

 const handlePageChange = (page) => {
  setSearchParams((prevParams) => {
   const params = new URLSearchParams(prevParams);
   params.set("page", page);
   return params;
  });
 };

 const handleItemsPerPageChange = (itemsPerPage) => {
  setSearchParams((prevParams) => {
   const params = new URLSearchParams(prevParams);
   params.set("page", 1);
   params.set("itemsPerPage", itemsPerPage);

   return params;
  });
 };

 const handleSortChange = (sort) => {
  setSearchParams((prevParams) => {
   const params = new URLSearchParams(prevParams);
   params.set("sort", sort);
   return params;
  });
 };

 useEffect(() => {
  const validCategory = categories.includes(category) ? category : "all";
  const validSort = [
   "name-asc",
   "name-desc",
   "price-asc",
   "price-desc",
  ].includes(sort)
   ? sort
   : "name-asc";
  const totalItems = displayedProducts.length;
  const maxPage = Math.ceil(totalItems / (itemPerPage || 10));
  const validPage = Math.max(1, Math.min(parseInt(page) || 1, maxPage));
  const validItemsPerPage = [5, 10, null].includes(parseInt(itemPerPage))
   ? itemPerPage
   : 10;

  const defaultParams = {
   category: validCategory,
   sort: validSort,
   page: validPage,
   itemsPerPage: validItemsPerPage,
   search: search,
  };

  const updatedParams = { ...defaultParams };

  if (
   JSON.stringify(updatedParams) !==
   JSON.stringify(Object.fromEntries(searchParams))
  ) {
   setSearchParams(updatedParams);
  }
 }, [
  categories,
  category,
  displayedProducts.length,
  itemPerPage,
  page,
  products,
  products.length,
  search,
  searchParams,
  setSearchParams,
  sort,
 ]);

 return (
  <>
   <Aside
    categories={categories}
    category={category}
    handleFilterChange={handleFilterChange}
   />
   <ShopMain
    category={category}
    products={displayedProducts}
    page={parseInt(page)}
    setPage={handlePageChange}
    itemPerPage={parseInt(itemPerPage)}
    setItemPerPage={handleItemsPerPageChange}
    sort={sort}
    setSort={handleSortChange}
    hoverButton={hoverButton}
    setHoverButton={setHoverButton}
    search={search}
   />
  </>
 );
}

export default Shop;
