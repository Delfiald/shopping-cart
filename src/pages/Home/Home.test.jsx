import { render, screen } from "@testing-library/react";
import Home from "./Home";
import { MemoryRouter, Outlet, Route, Routes } from "react-router-dom";
import userEvent from "@testing-library/user-event";
import Cart from "../Cart/Cart";
import { useState } from "react";
import PropTypes from "prop-types";
import { expect } from "vitest";
import Shop from "../Shop/Shop";

const MockRouter = ({
 context = {
  setHoverButton: () => {},
  setSearchInput: () => {},
 },
}) => {
 return (
  <MemoryRouter initialEntries={["/"]}>
   <Routes>
    <Route path="/" element={<Outlet context={context} />}>
     <Route index element={<Home />} />
     <Route path="cart" element={<Cart />} />
     <Route path="shop" element={<Shop />} />
    </Route>
   </Routes>
  </MemoryRouter>
 );
};

function HoverButtonTest({ cartItem, notificationItem }) {
 const [hoverButton, setHoverButton] = useState(null);

 return (
  <MockRouter
   context={{ hoverButton, setHoverButton, cartItem, notificationItem }}
  />
 );
}

function SearchInputTest() {
 const [searchInput, setSearchInput] = useState("");

 return (
  <MockRouter
   context={{ searchInput: searchInput, setSearchInput: setSearchInput }}
  />
 );
}

MockRouter.propTypes = {
 context: PropTypes.object,
};

HoverButtonTest.propTypes = {
 cartItem: PropTypes.array,
 notificationItem: PropTypes.array,
};

const mockCartItem = [
 { id: 1, description: "item1" },
 { id: 2, description: "item2" },
 { id: 3, description: "item3" },
];
const mockNotificationItem = [
 { id: 1, description: "message1" },
 { id: 2, description: "message2" },
];

describe("Home Page Test", () => {
 it("Renders Home Page", async () => {
  const { container } = render(<MockRouter />);
  expect(container).toMatchSnapshot();
 });

 it("Renders Header from Home Page", () => {
  render(<MockRouter />);

  const header = screen.getByRole("heading", { name: /shoppers/i });
  expect(header).toBeInTheDocument();
 });

 it("Renders search bar from Home Page", () => {
  render(<MockRouter />);

  const searchLabel = screen.getByTitle("Search Icon");
  expect(searchLabel).toBeInTheDocument();
  const search = screen.getByRole("textbox");
  expect(search).toBeInTheDocument();
 });

 it("Renders Cart Button", () => {
  render(<MockRouter />);

  const shoppingCart = screen.getByTestId("cart-button");
  expect(shoppingCart).toBeInTheDocument();

  const cartIcon = screen.getByTitle("Cart Button");
  expect(cartIcon).toBeInTheDocument();
 });

 it("Cart Button Routes to Cart Page when Clicked", async () => {
  const user = userEvent.setup();
  render(<MockRouter />);

  const shoppingCartButton = screen.getByTestId("cart-button");
  await user.click(shoppingCartButton);

  const cartHeading = screen.getByText("Cart");

  expect(cartHeading).toBeInTheDocument();

  const cartMain = screen.getByText("Cart Main");
  expect(cartMain).toBeInTheDocument();
 });

 it(`Doesn't Render Item Counter in Cart`, () => {
  render(<MockRouter context={{ cartItem: [] }} />);

  const itemCounter = screen.queryByTestId("item-count");
  expect(itemCounter).not.toBeInTheDocument();
 });

 it("Render Item Counter in Cart", () => {
  render(<MockRouter context={{ cartItem: mockCartItem }} />);

  const itemCounter = screen.getByTestId("item-count");
  expect(itemCounter).toBeInTheDocument();
  expect(itemCounter).toHaveTextContent(mockCartItem.length);
 });

 it("Open Cart when Hovered and Close it when not", async () => {
  const event = userEvent.setup();

  render(<HoverButtonTest cartItem={mockCartItem} />);

  const shoppingCartButton = screen.getByTestId("cart-button");

  await event.hover(shoppingCartButton);
  const cartItems = screen.getAllByText(/item/i);
  expect(cartItems).toHaveLength(mockCartItem.length);

  await event.unhover(shoppingCartButton);
  expect(screen.queryAllByText(/item/i)).toHaveLength(0);
 });

 it("Render Notifications Button", () => {
  render(<MockRouter />);

  const notificationButton = screen.getByTestId("notification-button");
  expect(notificationButton).toBeInTheDocument();
 });

 it(`Doesn't render notification Counter`, () => {
  render(<MockRouter context={{ notificationItem: [] }} />);

  const notificationCounter = screen.queryByTestId("notification-counter");
  expect(notificationCounter).not.toBeInTheDocument();
 });

 it("Render Notification Counter", () => {
  render(<MockRouter context={{ notificationItem: mockNotificationItem }} />);

  const notificationCounter = screen.getByTestId("notification-count");
  expect(notificationCounter).toBeInTheDocument();
  expect(notificationCounter).toHaveTextContent(mockNotificationItem.length);
 });

 it("Open notifications when hover and close it when not", async () => {
  const event = userEvent.setup();

  render(<HoverButtonTest notificationItem={mockNotificationItem} />);

  const notificationButton = screen.getByTestId("notification-button");

  await event.hover(notificationButton);
  const notificationItem = screen.getAllByText(/message/i);
  expect(notificationItem).toHaveLength(mockNotificationItem.length);
  await event.unhover(notificationButton);

  expect(screen.queryAllByText(/message/i)).toHaveLength(0);
 });

 it("Change Input Value for Search Box", async () => {
  const event = userEvent.setup();

  render(<SearchInputTest />);

  const searchInput = screen.getByRole("textbox");

  await event.type(searchInput, "Hello World");
  expect(searchInput).toHaveValue("Hello World");
 });

 it("Should Render Call to Action Section", async () => {
  const event = userEvent.setup();
  render(<MockRouter />);

  const ctaButton = screen.getByRole("button", { name: "Shop Now" });
  expect(ctaButton).toBeInTheDocument();

  await event.click(ctaButton);

  const shopMainHeading = screen.getByText("Shop Main");
  expect(shopMainHeading).toBeInTheDocument();
 });
});
