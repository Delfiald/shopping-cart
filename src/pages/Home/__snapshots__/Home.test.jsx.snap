// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Home Page Test > Renders Home Page 1`] = `
<div>
  <main
    class="_home_716e15"
  >
    <div
      class="_hero_716e15"
    >
      <div
        class="_carousel_716e15"
        data-testid="products-carousel-section"
      >
        <div
          class="_item-display_716e15 "
        >
          <div
            class="_displayed-image_716e15"
          >
            <img
              alt=""
              src=""
            />
          </div>
          <h2 />
        </div>
        <div
          class="_carousel-item-wrapper_716e15"
        >
          <div
            class="sc-KXCwU bGPffd _loading-bar_716e15"
          />
          <div
            class="_carousel-item_716e15 __716e15"
          >
            <div
              class="_image-wrapper_716e15"
            >
              <img
                alt="product-1"
                src="/image/product-1.png"
                title="product-1"
              />
            </div>
          </div>
          <div
            class="_carousel-item_716e15 __716e15"
          >
            <div
              class="_image-wrapper_716e15"
            >
              <img
                alt="product-2"
                src="/image/product-2.png"
                title="product-2"
              />
            </div>
          </div>
          <div
            class="_carousel-item_716e15 __716e15"
          >
            <div
              class="_image-wrapper_716e15"
            >
              <img
                alt="product-3"
                src="/image/product-3.png"
                title="product-3"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="_categories_716e15"
      data-testid="categories-section"
    >
      <h2>
        Categories
      </h2>
      <div
        class="_category_716e15"
      >
        <img
          alt="category-1"
          src="Categories/category-1.jpg"
        />
        <h3>
          category-1
        </h3>
      </div>
      <div
        class="_category_716e15"
      >
        <img
          alt="category-2"
          src="Categories/category-2.jpg"
        />
        <h3>
          category-2
        </h3>
      </div>
      <div
        class="_category_716e15"
      >
        <img
          alt="category-3"
          src="Categories/category-3.jpg"
        />
        <h3>
          category-3
        </h3>
      </div>
      <div
        class="_category_716e15"
      >
        <img
          alt="category-4"
          src="Categories/category-4.jpg"
        />
        <h3>
          category-4
        </h3>
      </div>
    </div>
    <div
      class="_call-to-action_716e15"
    >
      <div
        class="_cta-wrapper_716e15"
      >
        <div
          class="_premises-text_716e15"
        >
          <h2>
            Donâ€™t miss out on exclusive deals!
          </h2>
          <div>
            Tap the button and start your shopping journey
          </div>
        </div>
        <button
          class="_cta-button_716e15"
          data-testid="cta-button"
        >
          <div
            class="_displayed_716e15"
          >
            Shop Now
          </div>
          <div
            class="_hovered_716e15"
          >
            Shop Now
          </div>
        </button>
      </div>
      <div
        class="_cta-background-wrapper_716e15"
      >
        <img
          alt="Call to Action Background"
          src="/image/product-1.png"
        />
      </div>
    </div>
  </main>
</div>
`;
